Add SUPG to INS

lindsayad opened the pull request:
This adds consistent (omitting terms with second derivatives) streamline upwinding for incompressible Navier-Stokes. SUPG is a method that modifies the test function, essentially giving more weight to the upwind element. An optimal factor (alpha) for the upwinding is chosen such that at least for single component, one dimension scalar advection, nearly exact solutions at the nodes are obtained.

SUPG allows higher Reynolds and grid Peclet numbers to be simulated. An example of a simulation that will not converge at steady-state with standard Galerkin weighting, but does converge with PG weighting is shown below with comparison to an OpenFOAM simulation. The Reynolds number for this problem is 1000; the grid Peclet number is 25.

### moose

![moose_supg_lid_driven](https://user-images.githubusercontent.com/10248304/29196794-70be97e4-7dfc-11e7-935d-49ea7cb3c018.png)

### OpenFOAM

![foam_lid_driven](https://user-images.githubusercontent.com/10248304/29196798-77f97bf0-7dfc-11e7-9bde-a7ad5fb2bc9b.png)

Some other fun examples of the importance of consistent SUPG weighting (e.g. application of the PG perturbation to the test function to source terms and time derivatives) are shown below:

### Importance of consistent weighting of source terms

The "No SUPG" label here means that SUPG is applied only to the advection operator

![source_supg_effect_hill](https://user-images.githubusercontent.com/10248304/29196833-bbbb5dea-7dfc-11e7-9ee9-649864c892e0.png)

### Importance of consistent weighting of time derivative

![transient_supg_effect](https://user-images.githubusercontent.com/10248304/29196841-c71b6676-7dfc-11e7-83f9-49cab5ad8c81.png)

The latter two example problems are taken from Huges et. al. Computer Methods in Applied Mechanics and Engineering 32 (1982) 199-259.

Addresses #9643

lindsayad committed:
Add INSMomentumSUPG kernel (#9643)

lindsayad committed:
Fix up INSMomentumSUPG and add INSMomentumTimeDerivativeSUPG (#9643)

lindsayad committed:
Create test (#9643)

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/107096/) on 4585ef4 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9644/site)

lindsayad committed:
Add doc pages (#9643)

lindsayad commented:
Interest in this @jwpeterson ?

jwpeterson commented:
> Interest in this @jwpeterson?

Yes, let's look at this during the Tiger Team this week?

lindsayad committed:
Create jacobians for SUPG (#9643)

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/108240/) on ffde016 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 7bf8d809633be04489497db74a358342c87af567`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9644/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9644/style.patch | git apply -v`

lindsayad committed:
Add jacobian test (#9643)

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/108266/) on d7e454b wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 7bf8d809633be04489497db74a358342c87af567`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9644/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9644/style.patch | git apply -v`

jwpeterson commented on modules/navier_stokes/src/kernels/INSMomentumSUPG.C:
```diff
@@ -0,0 +1,326 @@
+  // derivative to the test function, which we are not equipped to do (?)
+  Real pressure_part = 0;
+  if (!_integrate_p_by_parts)
+    pressure_part = _grad_p[_qp](_component) * PG_test;
```
Per our discussion, I think the stabilization terms should involve grad(p) regardless of whether or not we integrate by parts.

lindsayad committed:
Make pressure term consistent (#9643)

jwpeterson commented on modules/navier_stokes/src/kernels/INSMomentumSUPG.C:
```diff
@@ -0,0 +1,312 @@
+
+  Real body_force_part = -_rho[_qp] * _gravity(_component) * PG_test;
+
+  return convective_part + pressure_part + /*viscous_part +*/ body_force_part;
```
As a longer term design goal, I think the "unstabilized" `INSMomentum` Kernel and the `INSMomentumSUPG` kernels should share a common base class which provides the strong (and weak) forms of the residual in a single location. That way, if new terms are added to the equation, they can be added in a single place rather than two different files. Code would also be able to be shared, since some parts of the strong and weak residual are the same (the parts that are not integrated by parts).

jwpeterson replied:
Not necessary to fix for this PR, but I'd like to know the tradeoffs involved in neglecting the viscous terms of the residual. Specifically, we should have some MMS tests to verify the convergence rates are still optimal when we make this error in consistency.

lindsayad replied:
Started working on PSPG (will be a different PR) and have created a base class like you suggested

jwpeterson commented on modules/navier_stokes/src/kernels/INSMomentumSUPG.C:
```diff
@@ -0,0 +1,312 @@
+    alpha = 1;
+  else if (U.norm() < std::numeric_limits<double>::epsilon())
+  {
+    alpha = 0;
```
Setting `alpha=0` is not necessary if you are just going to return.

jwpeterson commented on modules/navier_stokes/src/kernels/INSMomentumSUPG.C:
```diff
@@ -0,0 +1,312 @@
+    }
+    else if (U.norm() < std::numeric_limits<double>::epsilon())
+    {
+      alpha = 0;
```
redundant

jwpeterson commented on modules/navier_stokes/src/kernels/INSMomentumSUPG.C:
```diff
@@ -0,0 +1,312 @@
+    }
+    else if (U.norm() < std::numeric_limits<double>::epsilon())
+    {
+      alpha = 0;
```
redundant

jwpeterson commented on modules/navier_stokes/src/kernels/INSMomentumSUPG.C:
```diff
@@ -0,0 +1,312 @@
+        _current_elem->hmax() / 2. * _grad_test[_i][_qp] *
+        (d_alpha_d_vel_component * U / U.norm() +
+         alpha * (d_U_d_vel_component / U.norm() -
+                  U / (U.norm() * U.norm() * U.norm()) * _u_vel[_qp] * _phi[_j][_qp]));
```
The only difference between this line and the other components is `_u_vel` vs. `_v_vel` vs. `_w_vel`, right? Would it be possible to reduce some of the code duplication by mapping `jvar` into a `compnent` and using `U(component)` instead?

jwpeterson commented on modules/navier_stokes/tests/ins/jacobian_test/jacobian_supg_test.i:
```diff
@@ -0,0 +1,146 @@
+[]
+
+[Variables]
+  [./vel_x]
```
This test uses equal-order interpolation for velocity and pressure, so it is not LBB stable. Although it might work for the specific grid used in this test, you would potentially encounter issues if you try to refine the mesh, etc. I'd rather not have input files with unstable formulations in the module, because people create new simulations and tests by copying existing ones...

jwpeterson commented on modules/navier_stokes/tests/ins/lid_driven/lid_driven_supg.cmp:
```diff
@@ -0,0 +1,10 @@
+# No GLOBAL VARIABLES
+
+NODAL VARIABLES relative 1.e-6 floor 1.e-12
+	p      absolute 5.e-4
```
Seems like a very loose tolerance for pressure? Is this just copied from the other lid-driven custom comparison file?

jwpeterson reviewed (changes requested):
* Fix redundant alpha=0 settings
* Look into reducing code duplication in Jacobian calculations.
* Fix LBB-unstable Jacobian test.
* Double-check loose pressure tolerance on new lid-driven cavity test.

lindsayad committed:
Update tests to reflect now consistent (pressure term) stabilization

- Refine mesh to eliminate visual oscillations in pressure solution
- Decrease to one time step to keep test time reasonable
- use lu preconditioner for greatly improved linear iteration performance
- Add pinned pressure point to ensure reproducible solutions
- Tighten pressure comparison tolerance

Addresses #9643

lindsayad committed:
Reduce code duplication by adding computeDPGTest method (#9643)

lindsayad committed:
Move new test things into test/tests (#9643)

lindsayad committed:
Remove velocity_channel from root/tests (#9643)

lindsayad committed:
Make slow non-linear physics test heavy (#9643)
